### ISSUE TITLE
model trained is different with tree_method "gpu_hist" between on single GPU and on multi-GPU? 

### ISSUE THREAD

**Issue by joegaotao** [dmlc/xgboost#3887]
I tested `tree_method = 'gpu_hist'` on single GPU and multi-GPU, and I found the predictions are different. According to my understanding, the hist method first converts continuous feature values into
k discrete bins, and then construct a histogram with k bins for each feature. Single GPU or multi-GPU just distribute and evaluate the gradients and hess in k bins, so the trained model should be the same when controlling the randomness. Is there any mistake in my understanding?.

Test code in R (xgboost 0.81.0.1):

single GPU:
```
library('xgboost')
# Simulate N x p random matrix with some binomial response dependent on pp columns
set.seed(111)
N <- 1000000
p <- 50
pp <- 25
X <- matrix(runif(N * p), ncol = p)
betas <- 2 * runif(pp) - 1
sel <- sort(sample(p, pp))
m <- X[, sel] %*% betas - 1 + rnorm(N)
y <- rbinom(N, 1, plogis(m))

tr <- sample.int(N, N * 0.75)
dtrain <- xgb.DMatrix(X[tr,], label = y[tr])
dtest <- xgb.DMatrix(X[-tr,], label = y[-tr])
wl <- list(train = dtrain, test = dtest)

param <- list(objective = 'reg:logistic', eval_metric = 'auc', 
  subsample = 1, num_parallel_tree = 1, nthread = 1L, eta = 0.3, max_depth = 6,
  colsample_bytree = 1, max_bin = 64, seed = 2018, tree_method = 'gpu_hist', n_gpus = 1, gpu_id = 0,
  grow_policy = "lossguide", max_leaves = 1e3)
pt <- proc.time()
bst_gpu <- xgb.train(param, dtrain, watchlist = wl, nrounds = 20)
proc.time() - pt

test_pred <- predict(bst_gpu, newdata = X[-tr,])
head(test_pred)
[1] 0.4517163 0.2565421 0.5847197 0.2956713 0.7016808 0.4059679
```

multi-GPU:
```
library('xgboost')

# Simulate N x p random matrix with some binomial response dependent on pp columns
set.seed(111)
N <- 1000000
p <- 50
pp <- 25
X <- matrix(runif(N * p), ncol = p)
betas <- 2 * runif(pp) - 1
sel <- sort(sample(p, pp))
m <- X[, sel] %*% betas - 1 + rnorm(N)
y <- rbinom(N, 1, plogis(m))

tr <- sample.int(N, N * 0.75)
dtrain <- xgb.DMatrix(X[tr,], label = y[tr])
dtest <- xgb.DMatrix(X[-tr,], label = y[-tr])
wl <- list(train = dtrain, test = dtest)

param <- list(objective = 'reg:logistic', eval_metric = 'auc', 
  subsample = 1, num_parallel_tree = 1, nthread = 1L, eta = 0.3, max_depth = 6,
  colsample_bytree = 1, max_bin = 64, seed = 2018, tree_method = 'gpu_hist', n_gpus = 2, gpu_id = 1,
  grow_policy = "lossguide", max_leaves = 1e3)
  # max_depth = 20, grow_policy = "lossguide", max_leaves = 1e3)
pt <- proc.time()
bst_gpu <- xgb.train(param, dtrain, watchlist = wl, nrounds = 20)
proc.time() - pt

test_pred1 <- predict(bst_gpu, newdata = X[-tr,])
head(test_pred1)
[1] 0.5315008 0.2953689 0.6302567 0.2885332 0.6240512 0.3615361
```

**Comment by hcho3** [comment]
@joegaotao I think the two models are different to begin with. You should get text dump of each model and compare. The predictor itself should give the same answer for the same model, regardless of `n_gpus` set. Prediction is different because model trained is different.

**Comment by joegaotao** [comment]
@hcho3  I know the trained models  are different leading to the different prediction, but the seed, dataset and parameter are the same except the `n_gpu`. When specifying `tree_method="gpu_exact"`, the model is the same, while `tree_method="gpu_hist"`with different gpu number is different. So I wonder what it is the difference in multi-GPU and single GPU when implementing gpu_hist algorithm?

**Comment by hcho3** [comment]
@joegaotao GPU hist uses approximate quantile sketchs to produce candidate split points. The sketch algorithm is approximate and thus may produce different answers for different `n_gpus`. Do you see significant drop in AUC?

@RAMitchell @trivialfis Correct me if I'm wrong.

**Comment by joegaotao** [comment]
@hcho3 I use it in the regression mode with our customized evaluation metric, and I found the metric decayed significantly. The candidate split points is not  sketched by all data firstly?

**Comment by hcho3** [comment]
@joegaotao Does your example script also show drop in metric?

**Comment by trivialfis** [comment]
@hcho3 @joegaotao Let me take a look.  I didn't try to look into this is because reason described in #3889.  If this sorted out, the results from single GPU and multi-gpu won't be the same in future anyway.

**Comment by joegaotao** [comment]
@hcho3 the example script is just a toy example, the different predictions seems to have no effect in such scenario.

**Comment by joegaotao** [comment]
@trivialfis I see. If each device propose splits based on its subset of the data, the parameter `n_gpus` will introduce some randomness for the data split strategy...
